arch/x86/entry.S

```diff
@@ -173,9 +181,15 @@ ENTRY(enter_usermode)
     PUSHF
 
     /* Save stack pointer onto per-cpu */
-    mov %_ASM_DX, %gs:(usermode_private)
+    mov %_ASM_SP, %gs:usermode_private
 
-    syscall_to_usermode
+    /* provide sysenter the kernel stack to be used */
+    push %_ASM_DX
+    WRITE_MSR $MSR_SYSENTER_ESP, %_ASM_SP
+    pop %_ASM_DX
+
+    /* Move to user stack */
+    mov %_ASM_DX, %_ASM_SP
```

Comment:
We should keep the previous logic and order here. Just add the PUSH+WRMSR+POP after the PUSHF